src/psyclone/psyGen.py

```diff
@@ -1271,12 +1301,30 @@ def depth(self):
             my_depth += 1
         return my_depth
 
-    @abc.abstractmethod
-    def view(self, indent=0):
-        '''Abstract function to prints a text representation of the node.
+    def view(self, indent=0, index=None):
+        ''' Print out description of current node to stdout and
+        then call view() on all child nodes.
 
         :param int indent: depth of indent for output text.
+        :param int index: the position of this Node wrt its siblings or None.
+
         '''
+        # Those nodes in the PSyIR for which we do not prepend an index to any
```

Comment:
An alternative would be to make this general class recurse down with `entity.view(indent=indent + 1, index=None)` and re-implement it on `Schedule.view()` to recurse the next step down with the index number. Would that work?

Reply:
I like that idea and have tried it. I have to override the view() method in `ProfileNode`, `ExtractNode` and `Directive` (as well as `Schedule`) as they all have children that we want indexed. An alternative would be for me to introduce a new, `IndexedNode` class that overrides this view() method just once but that feels a bit heavyweight and might cause us problems with our class hierarchy - what do you think?

Reply:
Ah, I did not consider Profile/ExtractNode. But I see they are essentially blocks of code with a sequence of statement as children and some extra functionality. So why are they not just sub-classes of Schedule? Schedules are already the concept of a Node with an indexed sequence of children, right? This will give them the overloaded squares brackets as well which I think should be available to any Node that shows indexes in the view(). 

If that breaks to many things I am happy with the IndexedNode.

Reply:
I'm continuing this discussion in #507...